src/grav/test.py

```diff
@@ -0,0 +1,115 @@
+    # are all in one. -- wdella 2019-10
+    client = DockerClient("/var/run/docker.sock")
+
+    # load image from tar, get image id
+    with open(image_tar, 'rb') as fh:
+        data = fh.read()
+
+    headers = {'Content-Type': 'application/tar', 'Content-Length': len(data)}
+    resp = client.post("/images/load?fromSrc=-", body=data, headers=headers)
+    assert resp.status in (200, 201)
+    json = resp.json()
+    match = re.match(r"Loaded image ID: (sha256:[\da-f]{64})", json['stream'])
+    image_id = match.group(1)
+
+    # create container, get container id
+    data = {"Image": image_id}
+    headers = {"Content-Type": "application/json"}
+    resp = client.post("/containers/create", body=data, headers=headers)
+    assert resp.status in (200, 201)
+    json = resp.json()
+    container_id = json['Id']
+
+    return client, image_id, container_id
+
+
+def test(client, container_id):
+    # start
+    start_path = "/containers/%s/start" % container_id
+    resp = client.post(start_path)
+    resp.read()  # TODO: have this handled by the client
+    assert resp.status == 204
+
+    # stop
+    stop_path = "/containers/%s/stop" % container_id
+    resp = client.post(stop_path)
+    resp.read()  # TODO: have this handled by the client
+    assert resp.status == 204
+
+    # logs
+    logs_path = "/containers/%s/logs" % container_id
+    logs_path += "?stdout=true&stderr=true"
+    resp = client.get(logs_path)
+    assert resp.status in (200, 201)
+    # logs doesn't return json, but a multiplexed stream
+    stdout, stderr = demux_logs(resp.read())
+
+    # validate
+    lines = stdout.splitlines()
+    assert lines[0] == "Hello Gravitational!"
+    assert lines[-1] == "Terminated. Bye!"
+    assert stderr == ""
+
+
+def teardown(client, container_id, image_id):
+    # if the test dies partway through, the container may
+    # still be running, and should be stopped
+    inspect_path = "/containers/%s/json" % container_id
+    resp = client.get(inspect_path)
+    assert resp.status == 200
+    json = resp.json()
+    state = json["State"]
+    running = state["Running"]
+    if running:
+        # stop
+        stop_path = "/containers/%s/stop" % container_id
+        resp = client.post(stop_path)
+        resp.read()  # TODO: have this handled by the client
+        assert resp.status == 204
+
+    # delete container
+    delete_path = "/containers/" + container_id
+    resp = client.request("DELETE", delete_path)
+    resp.read()  # TODO: have this handled by the client
+    assert resp.status == 204
+
+    # delete image
+    delete_path = "/images/" + image_id
+    resp = client.request("DELETE", delete_path)
+    resp.read()  # TODO: have this handled by the client
+    assert resp.status == 200  # why isn't this 204 like the other api endpoints?
+
+
+if __name__ == "__main__":
+    client, image_id, container_id = setup()
+    test(client, container_id)
+    teardown(client, container_id, image_id)
```

Comment:
teardown can be a __finally__ to clean up in case of exceptions?

Reply:
Per my remarks at the top of the PR: I intentionally left out error handling, logging/output, and anything related to the test harness to keep PR review more manageable -- those will come in a later PR.

Let me know if you consider this a blocker for this code.  Otherwise, I'll file an issue for it and address it in a future PR.

Reply:
Tracked in https://github.com/javins/gravitational-audition/issues/8

Reply:
Addressed by ec8d565  in PR #9